### PR TITLE
feat: generate pr description from configuration comparison result

### PR DESCRIPTION
**Pull request by JoeWang1127**
In this PR:
- Generate pull request description from configuration comparison result. 
- The pull request description contains repo-level changes, if applicable, and googleapis commit from baseline config (exclusive) to current config (inclusive).

An example of pr description with repo-level change:
```
This pull request is generated with proto changes between [googleapis/googleapis@3b6f144](https://github.com/googleapis/googleapis/commit/3b6f144d47b0a1d2115ab2445ec06e80cc324a44) (exclusive) and [googleapis/googleapis@0cea717](https://github.com/googleapis/googleapis/commit/0cea7170404bec3d994f43db4fa292f5034cbe9a) (inclusive).

BEGIN_COMMIT_OVERRIDE
BEGIN_NESTED_COMMIT
fix(deps): update the Java code generator (gapic-generator-java) to 1.2.3
END_NESTED_COMMIT
BEGIN_NESTED_COMMIT
chore: update the libraries_bom version to 2.3.4
END_NESTED_COMMIT
BEGIN_NESTED_COMMIT
feat: Make Layout Parser generally available in V1

PiperOrigin-RevId: 638924855

Source Link: [googleapis/googleapis@0cea717](https://github.com/googleapis/googleapis/commit/0cea7170404bec3d994f43db4fa292f5034cbe9a)
END_NESTED_COMMIT
END_COMMIT_OVERRIDE
```

### CODE REVIEW

File: library_generation/test/generate_pr_description_unit_tests.py [hunk 1]
```diff
@@ -38,36 +49,41 @@ def test_get_commit_messages_current_is_older_raise_exception(self):
             True,
         )
 
-    def test_get_commit_messages_current_and_baseline_are_same_raise_exception(self):
+    def test_get_commit_messages_current_and_baseline_are_returns_empty_message(self):
```
Comment (diegomarquezp) [inline library_generation/test/generate_pr_description_unit_tests.py:52]: nit: Shouldn't this be "get commit messages with _equal/same_ current and baseline commit"?

File: library_generation/test/generate_pr_description_unit_tests.py [hunk 2]
```diff
@@ -38,36 +49,41 @@ def test_get_commit_messages_current_is_older_raise_exception(self):
...
-        cwd = os.getcwd()
+        config = GenerationConfig(
+            gapic_generator_version="",
+            googleapis_commitish=commit_sha,
+            libraries_bom_version="",
+            # use empty libraries to make sure no qualified commit between
+            # two commit sha.
+            libraries=[],
+        )
+        path = os.path.join(os.getcwd(), "no_config_change")
```
Comment (diegomarquezp) [inline library_generation/test/generate_pr_description_unit_tests.py:77]: nit: `path` could have a more descriptive name, like `pr_description_path`

File: library_generation/test/generate_pr_description_unit_tests.py [hunk 3]
```diff
@@ -76,19 +92,81 @@ def test_generate_pr_description_does_not_create_pr_description_without_qualifie
         old_commit_sha = "30717c0b0c9966906880703208a4c820411565c4"
         # committed on May 23rd, 2024
         new_commit_sha = "eeed69d446a90eb4a4a2d1762c49d637075390c1"
+        path = os.path.join(os.getcwd(), "no_qualified_commit")
```
Comment (diegomarquezp) [inline library_generation/test/generate_pr_description_unit_tests.py:95]: nit: same comment here about the variable name.

File: library_generation/generate_pr_description.py [hunk 4]
```diff
@@ -17,54 +17,51 @@
...
-    config: GenerationConfig,
-    baseline_commit: str,
+    config_change: ConfigChange,
     description_path: str,
     repo_url: str = "https://github.com/googleapis/googleapis.git",
 ) -> None:
     """
-    Generate pull request description from baseline_commit (exclusive) to the
-    googleapis commit (inclusive) in the given generation config.
+    Generate pull request description from configuration comparison result.
```
Comment (diegomarquezp) [inline library_generation/generate_pr_description.py:38]: This comment change nicely describes how the pr_description generation will be changed after merging this PR. Could you please add a similar note in the PR description? I could not immediately understand what this PR is about without reading this comment and the implementation changes in `commit_mesage_formatter.py`.

**Review by diegomarquezp — APPROVED** [review]
Thanks for the PR. Just a few nits + 1 comment.

File: library_generation/generate_pr_description.py (hunk 4 above)
Comment (JoeWang1127) [reply library_generation/generate_pr_description.py:38]: Done.

File: library_generation/test/generate_pr_description_unit_tests.py (hunk 2 above)
Comment (JoeWang1127) [reply library_generation/test/generate_pr_description_unit_tests.py]: Done.

File: library_generation/generate_pr_description.py [hunk 5]
```diff
@@ -97,10 +95,13 @@ def get_commit_messages(
     selecting commit message. This commit should be an ancestor of
     :param paths: a mapping from file paths to library_name.
     :param is_monorepo: whether to generate commit messages in a monorepo.
+    :param repo_level_message: commit messages regarding repo-level changes.
     :return: commit messages.
     :raise ValueError: if current_commit is older than or equal to
     baseline_commit.
     """
+    if current_commit_sha == baseline_commit_sha:
+        return EMPTY_MESSAGE
```
Comment (blakeli0) [inline library_generation/generate_pr_description.py:104]: The change could still come from library level change(e.g. api_description) even with the same commit_sha. Ideally we should combine all levels of changes, but I know this function might be only for generating repo level commit messages. If that's the case, can we rename the function?

File: library_generation/generate_pr_description.py [hunk 6]
```diff
@@ -97,10 +95,13 @@ def get_commit_messages(
     selecting commit message. This commit should be an ancestor of
     :param paths: a mapping from file paths to library_name.
     :param is_monorepo: whether to generate commit messages in a monorepo.
+    :param repo_level_message: commit messages regarding repo-level changes.
     :return: commit messages.
     :raise ValueError: if current_commit is older than or equal to
     baseline_commit.
     """
+    if current_commit_sha == baseline_commit_sha:
```
Comment (blakeli0) [inline library_generation/generate_pr_description.py:103]: What is the behavior if current_commit_sha < baseline_commit_sha? It is not a valid scenario now, but maybe we need to revert the changes in some cases. The scripts should at least not blow up at this point.

File: library_generation/test/resources/goldens/pr_description-golden.txt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+BEGIN_COMMIT_OVERRIDE
+BEGIN_NESTED_COMMIT
+Update repo-level parameter gapic_generator_version to 1.2.3
```
Comment (blakeli0) [inline library_generation/test/resources/goldens/pr_description-golden.txt:3]: Can we change this to something like `fix(deps): Update the Java code generator (gapic-generator-java) to 2.39.0` to keep it consistent with the existing Owlbot behavior? See https://github.com/googleapis/java-bigtable/pull/2191

File: library_generation/generate_pr_description.py [hunk 8]
```diff
@@ -156,24 +156,13 @@ def __filter_qualified_commit(paths: Dict[str, str], commit: Commit) -> (Commit,
 
 
 def __combine_commit_messages(
-    current_commit: Commit,
-    baseline_commit: Commit,
     commits: Dict[Commit, str],
     is_monorepo: bool,
+    repo_level_message: list[str],
 ) -> str:
-    messages = [
```
Comment (blakeli0) [inline library_generation/generate_pr_description.py:164]: This is pretty useful to handwritten library owners, they even proposed to enhance it to a diff. We don't have to change anything, but let's still keep it if there is a googleapis_commitish change.

File: library_generation/test/resources/goldens/pr_description-golden.txt [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+END_NESTED_COMMIT
+BEGIN_NESTED_COMMIT
+Update repo-level parameter libraries_bom_version to 2.3.4
+END_NESTED_COMMIT
+BEGIN_NESTED_COMMIT
+Update repo-level parameter protoc_version to 3.4.5
```
Comment (blakeli0) [inline library_generation/test/resources/goldens/pr_description-golden.txt:9]: We should remove protoc_version from our test proto now.

File: library_generation/test/resources/goldens/pr_description-golden.txt [hunk 10]
```diff
@@ -0,0 +1,18 @@
+BEGIN_COMMIT_OVERRIDE
+BEGIN_NESTED_COMMIT
+Update repo-level parameter gapic_generator_version to 1.2.3
+END_NESTED_COMMIT
+BEGIN_NESTED_COMMIT
+Update repo-level parameter libraries_bom_version to 2.3.4
```
Comment (blakeli0) [inline library_generation/test/resources/goldens/pr_description-golden.txt:6]: Can we add a conventional commit prefix? This could be a chore IMO.

File: library_generation/test/resources/goldens/pr_description-golden.txt (hunk 7 above)
Comment (JoeWang1127) [reply library_generation/test/resources/goldens/pr_description-golden.txt:3]: We need a mapping to do this, from parameter name to description.

What's the benefits of doing this? We don't have to generate the same message.

File: library_generation/generate_pr_description.py (hunk 6 above)
Comment (JoeWang1127) [reply library_generation/generate_pr_description.py:103]: We can't compare different commit sha directly. We get the commit time of two commits and if the current commit is older than baseline commit, the script will exit with error (we have unit test to verify this behavior).

File: library_generation/generate_pr_description.py (hunk 5 above)
Comment (JoeWang1127) [reply library_generation/generate_pr_description.py:104]: Library level change will not generate a pr description since we have a dedicated workflow to generate the description.

Comment (JoeWang1127) [reply library_generation/generate_pr_description.py:104]: Done.

File: library_generation/generate_pr_description.py (hunk 8 above)
Comment (JoeWang1127) [reply library_generation/generate_pr_description.py:164]: I restored this change.

File: library_generation/test/resources/goldens/pr_description-golden.txt (hunk 7 above)
Comment (blakeli0) [reply library_generation/test/resources/goldens/pr_description-golden.txt]: > What's the benefits of doing this? I think we don't have to generate the same message as long as the meaning of this message is clear.

I agree. But I think the previous message makes more sense for handwritten library owners. Because
1. It includes a conventional commit prefix.
2. They don't need to know something like `repo-level parameter`, it is an implementation detail of hermetic build scripts. On high level, they just need to know the generator version is updated and the GAPIC layer is regenerated.

File: library_generation/generate_pr_description.py (hunk 6 above)
Comment (blakeli0) [reply library_generation/generate_pr_description.py:104]: > the script will exit with error

I remember it now. Thanks!

File: library_generation/test/resources/goldens/pr_description-golden.txt (hunk 9 above)
Comment (blakeli0) [reply library_generation/test/resources/goldens/pr_description-golden.txt]: Yes, but using protoc_version might also confuse other people. Can you use googleapis_commitish instead? Or use something random to indicate that the current logic can be used for any repo level changes?

Comment (JoeWang1127) [reply library_generation/test/resources/goldens/pr_description-golden.txt]: > Can you use googleapis_commitish instead?

googleapis commit will not appear here because the commits are formatted to separate messages.

I'll remove this message.

Comment (JoeWang1127) [reply library_generation/test/resources/goldens/pr_description-golden.txt]: Done

File: library_generation/test/resources/goldens/pr_description-golden.txt (hunk 7 above)
Comment (JoeWang1127) [reply library_generation/test/resources/goldens/pr_description-golden.txt]: Sure, this does make sense.

I need to create a mapping from parameter to commit message and I'll only add generator version and libraries_bom version for now. Other parameter will fallback to default message.

Comment (JoeWang1127) [reply library_generation/test/resources/goldens/pr_description-golden.txt]: Done.

**Comment by JoeWang1127** [comment]
Downstream check failed in java-spanner-jdbc, but I don't think this related to this PR.
```
Setup maven mirror
Installing this repo's modules to local maven.
Install complete. java-shared-dependencies = 3.31.1-SNAPSHOT
~/work/sdk-platform-java/sdk-platform-java/java-shared-dependencies/target ~/work/sdk-platform-java/sdk-platform-java
Cloning into 'java-spanner-jdbc'...
warning: Could not find remote branch v2.1[9](https://github.com/googleapis/sdk-platform-java/actions/runs/9453493033/job/26039662521?pr=2841#step:7:10).2 to clone.
fatal: Remote branch v2.19.2 not found in upstream origin
```

**Review by blakeli0 — APPROVED** [review]
